scripts.py

```diff
@@ -2497,16 +2500,30 @@ def update_timestamps(self, migration_file):
                 self._db, finish=last_run_date,
                 counter=counter, migration_name=migration_file
             )
+        
+        # Nothing to update
+        if self.overall_timestamp is None:
+            return
 
-        if (self.overall_timestamp and
-            (self.overall_timestamp.finish < last_run_date or
-            (self.overall_timestamp.finish==last_run_date and
-             self.overall_timestamp.counter < counter))
-        ):
-            self.overall_timestamp.update(
-                self._db, finish=last_run_date,
-                counter=counter, migration_name=migration_file
-            )
+        # The last script that ran had an earlier timestamp than the current script
+        if self.overall_timestamp.finish > last_run_date:
+            return
+
+        # The dates of the scrips are the same so compare the counters
+        if self.overall_timestamp.finish==last_run_date:
+            # The current script has no counter, so it's the same script that ran
+            # or an earlier script that ran
+            if counter is None:
+                return
+            # The previous script has a higher counter
+            if (self.overall_timestamp.counter is not None and
+                self.overall_timestamp.counter > counter):
+                return
+
+        self.overall_timestamp.update(
+            self._db, finish=last_run_date,
+            counter=counter, migration_name=migration_file
+        )
```

Comment:
This is an especially bad case of "can't compare None against a non-None value". We had to unroll the entire algorithm.